constraint-solver/src/solver/linearized.rs

```diff
@@ -0,0 +1,449 @@
+use std::iter;
+use std::{collections::HashSet, fmt::Display};
+
+use itertools::Itertools;
+
+use crate::constraint_system::ConstraintSystem;
+use crate::indexed_constraint_system::apply_substitutions;
+use crate::runtime_constant::Substitutable;
+use crate::solver::var_transformation::Variable;
+use crate::solver::{Error, VariableAssignment};
+use crate::{
+    constraint_system::BusInteraction,
+    grouped_expression::{GroupedExpression, RangeConstraintProvider},
+    range_constraint::RangeConstraint,
+    runtime_constant::{RuntimeConstant, VarTransformable},
+    solver::Solver,
+};
+
+/// A Solver that turns algebraic constraints into affine constraints
+/// by introducing new variables for the non-affine parts.
+/// It also replaces bus interaction fields by new variables if they are
+/// not just variables or constants.
+///
+/// The original algebraic constraints are kept as well.
+pub struct LinearizedSolver<T, V, S> {
+    solver: S,
+    linearizer: Linearizer<T, V>,
+    var_dispenser: LinearizedVarDispenser,
+}
+
+struct LinearizedVarDispenser {
+    next_var_id: usize,
+}
+
+impl LinearizedVarDispenser {
+    fn new() -> Self {
+        LinearizedVarDispenser { next_var_id: 0 }
+    }
+
+    fn next_var<V>(&mut self) -> Variable<V> {
+        let id = self.next_var_id;
+        self.next_var_id += 1;
+        Variable::Linearized(id)
+    }
+
+    /// Returns an iterator over all variables dispensed in the past.
+    fn all_dispensed_vars<V>(&self) -> impl Iterator<Item = Variable<V>> {
+        (0..self.next_var_id).map(Variable::Linearized)
+    }
+}
+
+impl<T, V, S> LinearizedSolver<T, V, S>
+where
+    T: RuntimeConstant,
+    V: Clone + Eq,
+    S: Solver<T, V>,
+{
+    pub fn new(solver: S) -> Self {
+        Self {
+            solver,
+            linearizer: Linearizer::default(),
+            var_dispenser: LinearizedVarDispenser::new(),
+        }
+    }
+}
+
+impl<T, V, S> RangeConstraintProvider<T::FieldType, Variable<V>>
+    for LinearizedSolver<T, Variable<V>, S>
+where
+    T: RuntimeConstant,
+    S: RangeConstraintProvider<T::FieldType, Variable<V>>,
+    V: Clone,
+{
+    fn get(&self, var: &Variable<V>) -> RangeConstraint<T::FieldType> {
+        self.solver.get(var)
+    }
+}
+
+impl<T: VarTransformable<V, Variable<V>>, V, S: Display> Display
+    for LinearizedSolver<T, Variable<V>, S>
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.solver)
+    }
+}
+
+impl<T, V, S> Solver<T, Variable<V>> for LinearizedSolver<T, Variable<V>, S>
+where
+    T: RuntimeConstant + Substitutable<Variable<V>> + Display + Hash,
+    V: Ord + Clone + Eq + Hash + Display,
+    S: Solver<T, Variable<V>>,
+{
+    fn solve(&mut self) -> Result<Vec<VariableAssignment<T, Variable<V>>>, Error> {
+        assert!(self.linearizer.constraints_to_add.is_empty());
+        let assignments = self.solver.solve()?;
+        self.linearizer.apply_assignments(&assignments);
```

Comment:
Do I understand this right:
- `self.solver.solve()?`mutates the underlying constraints, replacing some expressions by other expressions (mostly single variables by constants)
- We expect that the caller applies these assignments, so any constraint added in the future, or query to `are_expressions_known_to_be_different`, etc would have these assignments applied

And that's why we want to mutate the substitution in the linearizer?

I think this deserves a comment :)